File: source/WeddingManager.Web/js/services/serviceDetailService.js
(function(){
    angular.module('weddingManager')
    .service('serviceDetailService', function(serviceDetailFactory){
        var output = {};
        var _serviceDetails = [];
        
        output.createServiceDetail = function(serviceId, onSuccess, onError){
            var serviceDetail = {
                Details : "New service detail",
                Location : "",
                StartTime : new Date()
            };
            serviceDetailFactory.createServiceDetail(serviceId, serviceDetail, function(serviceDetailId){
                serviceDetail.Id = serviceDetailId;
                onSuccess(serviceDetail);
            }, onError);
        }
        
        output.refreshServiceDetails = function(serviceId, onSuccess, onError){
            serviceDetailFactory.retrieveServiceDetails(serviceId, function(serviceDetails){
                var serviceDetailCount = serviceDetails.length;
                for(var index = 0; index < serviceDetailCount; index++){
                    var serviceDetail = serviceDetails[index];
                    serviceDetail.StartTime = new Date(serviceDetail.StartTime);
                    if(serviceDetail.EndTime){
                        serviceDetail.EndTime = new Date(serviceDetail.EndTime);
                    }else{
                        serviceDetail.EndTime = null;
                    }
                }
                _serviceDetails = serviceDetails;
                onSuccess(_serviceDetails);
            }, onError);
        }
        
        output.updateServiceDetail = function(serviceDetail, onSuccess, onError){
            serviceDetailFactory.updateServiceDetail(serviceDetail, function(){
                onSuccess();
            }, onError);
        }
        
        output.deleteServiceDetail = function(serviceDetailId, onSuccess, onError){
            serviceDetailFactory.deleteServiceDetail(serviceDetailId, function(){
                onSuccess();
            }, onError);
        }
        
        output.search = function(companyId, searchCriteria, onSuccess, onError){
            serviceDetailFactory.search(companyId, searchCriteria, function(searchResults){
                onSuccess(searchResults);
            }, onError);
        }
        
        return output;
    });
})();

File: source/WeddingManager.Services/ServiceDetailService.cs
using System.Collections.Generic;
using WeddingManager.Core.Data;
using WeddingManager.Core.Repositories;
using WeddingManager.Core.Services;

namespace WeddingManager.Services
{
    public class ServiceDetailService : IServiceDetailService
    {
        public ServiceDetailService(IServiceDetailRepository serviceDetailRepository)
        {
            _serviceDetailRepository = serviceDetailRepository;
        }

        public int CreateServiceDetail(int serviceId, ServiceDetail serviceDetail)
        {
            return _serviceDetailRepository.CreateServiceDetail(serviceId, serviceDetail);
        }

        public IEnumerable<ServiceDetail> RetrieveServiceDetails(int serviceId)
        {
            return _serviceDetailRepository.RetrieveServiceDetails(serviceId);
        }

        public void UpdateServiceDetail(ServiceDetail serviceDetail)
        {
            _serviceDetailRepository.UpdateServiceDetail(serviceDetail);
        }

        public void DeleteServiceDetail(int serviceDetailId)
        {
            _serviceDetailRepository.DeleteServiceDetail(serviceDetailId);
        }

        public IEnumerable<ServiceDetailSearchResult> Search(int companyId, ServiceDetailSearchCriteria searchCriteria)
        {
            return _serviceDetailRepository.Search(companyId, searchCriteria);
        }

        private readonly IServiceDetailRepository _serviceDetailRepository;
    }
}